polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -338,12 +357,183 @@ static ValueOrMultiplier getMultiplier(const Value expr, const Value factor,
+                isZero(lhs.get<Value>(), solver))
+              return rhs;
+          }
+
+          // Otherwise, if neither the LHS nor the RHS subtrees passed up an
+          // offset, the offset is the operand that does not use any loop IV or
+          // thread id.
+          if (!lhsIsOff && !rhsIsOff) {
+            if (usesAny(lhs.get<Value>(), loopAndThreadVars) &&
+                !usesAny(rhs.get<Value>(), loopAndThreadVars))
+              return Offset(rhs.get<Value>());
+            if (usesAny(rhs.get<Value>(), loopAndThreadVars) &&
+                !usesAny(lhs.get<Value>(), loopAndThreadVars))
+              return Offset(lhs.get<Value>());
+          }
+
+          return Value();
+        };
+
+        LLVM_DEBUG(llvm::dbgs() << "expr: " << expr << "\n");
+
+        return visitBinaryOp(binOp, loopAndThreadVars, solver, getOffset,
+                             computeResult);
+      })
+      .Case<arith::MulIOp>([&](auto mulOp) -> ValueOr<Offset> {
+        auto computeResult = [&](ValueOr<Offset> lhs,
+                                 ValueOr<Offset> rhs) -> ValueOr<Offset> {
+          bool lhsIsOff = lhs.is<Offset>(), rhsIsOff = rhs.is<Offset>();
+
+          // If both the LHS and RHS subtrees passed up an offset, and either is
+          // one, return the other.
+          if (lhsIsOff && rhsIsOff) {
+            if (lhs.isEqualTo(1, solver))
+              return rhs;
+            if (rhs.isEqualTo(1, solver))
+              return lhs;
+          }
+
+          // If the LHS (or RHS) subtrees passed up an offset. Return:
+          //   - the offset if the other operand has value one
+          //   - a new offset with value equal to the other operand if the
+          //     offset is one
+          if (lhsIsOff && !rhsIsOff) {
+            if (lhs.isEqualTo(0, solver) ||
+                isEqualTo(rhs.get<Value>(), 1, solver))
+              return lhs;
+            if (lhs.isEqualTo(1, solver))
+              return Offset(rhs.get<Value>());
+          }
+          if (rhsIsOff && !lhsIsOff) {
+            if (rhs.isEqualTo(0, solver) ||
+                isEqualTo(lhs.get<Value>(), 1, solver))
+              return rhs;
+            if (rhs.isEqualTo(1, solver))
+              return Offset(lhs.get<Value>());
+          }
+
+          // Otherwise, if neither the LHS nor the RHS subtrees passed up an
+          // offset, and the multiplication uses a loop IV or thread id, the
+          // offset is zero.
+          if (!lhsIsOff && !rhsIsOff && usesAny(mulOp, loopAndThreadVars))
+            return Offset::zero(expr.getContext());
+
+          return Value();
+        };
+
+        LLVM_DEBUG(llvm::dbgs() << "expr: " << expr << "\n");
+
+        return visitBinaryOp(mulOp, loopAndThreadVars, solver, getOffset,
+                             computeResult);
+      })
+      .Case<CastOpInterface>([&](auto) {
+        return getOffset(op->getOperand(0), loopAndThreadVars, solver);
```

Comment:
return `Value()` or `expr` when number of operands != 1, as operation with CastOpInterface can have N inputs and N outputs.

Reply:
I guess there isn't anything that prevents an operation that implements the `castOpInterface` to have more than a single operand, although that would be strange. I will add an assertion.